File: lib/services/lyric_romanizer/engines/thai_engine.dart
/// Thai to Latin romanization engine.
///
/// Uses a character-mapping approach based on RTGS
/// (Royal Thai General System of Transcription).

library;
import '../lyric_romanizer_types.dart';

class ThaiRomanizer {
  const ThaiRomanizer();

  static const Map<String, String> _consonants = {
    'ก': 'k', 'ข': 'kh', 'ค': 'kh', 'ง': 'ng', 'จ': 'ch', 'ฉ': 'ch',
    'ช': 'ch', 'ซ': 's', 'ด': 'd', 'ต': 't', 'ถ': 'th', 'ท': 'th',
    'น': 'n', 'บ': 'b', 'ป': 'p', 'ผ': 'ph', 'ฝ': 'f', 'พ': 'ph',
    'ฟ': 'f', 'ม': 'm', 'ย': 'y', 'ร': 'r', 'ล': 'l', 'ว': 'w',
    'ส': 's', 'ห': 'h', 'อ': '', 'ฮ': 'h',
  };

  static const Map<String, String> _vowels = {
    'ะ': 'a', 'า': 'a', 'ิ': 'i', 'ี': 'i', 'ึ': 'ue', 'ื': 'ue',
    'ุ': 'u', 'ู': 'u', 'เ': 'e', 'แ': 'ae', 'โ': 'o', 'ใ': 'ai',
    'ไ': 'ai', 'ๅ': 'aa',
  };

  static const Map<String, String> _tones = {
    '่': '', '้': '', '๊': '', '๋': '',
  };

  static const Map<String, String> _combining = {
    'ำ': 'am', 'ๅ': 'aa',
  };

  Future<String> romanize(String line, RomanizeEngineContext context) async {
    final buf = StringBuffer();
    int i = 0;
    while (i < line.length) {
      final ch = line[i];
      if (_combining.containsKey(ch)) {
        buf.write(_combining[ch]);
        i++;
        continue;
      }
      if (_consonants.containsKey(ch)) {
        buf.write(_consonants[ch]);
        i++;
        continue;
      }
      if (_vowels.containsKey(ch)) {
        buf.write(_vowels[ch]);
        i++;
        continue;
      }
      if (_tones.containsKey(ch)) {
        // Tone marks are silent in RTGS
        i++;
        continue;
      }
      buf.write(ch);
      i++;
    }
    return buf.toString();
  }
}

File: android/app/src/main/kotlin/com/nomadguy/noctra/lyricromanizer/engines/ThaiEngine.kt
package com.nomadguy.noctra.lyricromanizer.engines

import com.nomadguy.noctra.lyricromanizer.RomanizeEngineContext

/** Thai to Latin romanizer (RTGS-based). */
class ThaiEngine {

    private val consonants = mapOf(
        'ก' to "k", 'ข' to "kh", 'ค' to "kh", 'ง' to "ng", 'จ' to "ch",
        'ฉ' to "ch", 'ช' to "ch", 'ซ' to "s", 'ด' to "d", 'ต' to "t",
        'ถ' to "th", 'ท' to "th", 'น' to "n", 'บ' to "b", 'ป' to "p",
        'ผ' to "ph", 'ฝ' to "f", 'พ' to "ph", 'ฟ' to "f", 'ม' to "m",
        'ย' to "y", 'ร' to "r", 'ล' to "l", 'ว' to "w", 'ส' to "s",
        'ห' to "h", 'อ' to "", 'ฮ' to "h"
    )

    private val vowels = mapOf(
        'ะ' to "a", 'า' to "a", 'ิ' to "i", 'ี' to "i",
        'ึ' to "ue", 'ื' to "ue", 'ุ' to "u", 'ู' to "u",
        'เ' to "e", 'แ' to "ae", 'โ' to "o", 'ใ' to "ai", 'ไ' to "ai"
    )

    private val toneMarks = setOf('่', '้', '๊', '๋')

    private val combining = mapOf('ำ' to "am")

    suspend fun romanize(line: String, context: RomanizeEngineContext): String {
        val buf = StringBuilder()
        var i = 0
        while (i < line.length) {
            val ch = line[i]
            val comb = combining[ch]
            if (comb != null) {
                buf.append(comb)
                i++; continue
            }
            val cons = consonants[ch]
            if (cons != null) {
                buf.append(cons)
                i++; continue
            }
            val vow = vowels[ch]
            if (vow != null) {
                buf.append(vow)
                i++; continue
            }
            if (ch in toneMarks) {
                i++; continue
            }
            buf.append(ch)
            i++
        }
        return buf.toString()
    }
}
